### PR TITLE
CMSIS-Core(M): Armv8-M Secure Stack Sealing

### PR DESCRIPTION
**Pull request by GuentherMartin**
- updated gcc, armclang CMSIS header files.
 - updated ARM, GCC startup files and linker description / scatter files.
 - configured C-Startup as default.

### CODE REVIEW

File: Device/ARM/ARMCM23/Source/ARM/ARMCM23_ac6.sct [hunk 1]
```diff
@@ -32,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM23/Source/ARM/ARMCM23_ac6.sct:52]: I believe it should be 'boundary '.

File: Device/ARM/ARMCM23/Source/ARM/ARMCM23_ac6_s.sct [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM23/Source/ARM/ARMCM23_ac6_s.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM23/Source/ARM/ARMCM23_ac6.sct [hunk 3]
```diff
@@ -32,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
 
-#define __RW_BASE       __RAM_BASE
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM23/Source/ARM/ARMCM23_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM33/Source/ARM/ARMCM33_ac6.sct [hunk 4]
```diff
@@ -32,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM33/Source/ARM/ARMCM33_ac6.sct:52]: I believe it should be 'boundary ',

File: Device/ARM/ARMCM33/Source/ARM/ARMCM33_ac6.sct [hunk 5]
```diff
@@ -32,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
 
-#define __RW_BASE       __RAM_BASE
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM33/Source/ARM/ARMCM33_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM33/Source/ARM/ARMCM33_ac6_s.sct [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM33/Source/ARM/ARMCM33_ac6_s.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM33/Source/ARM/startup_ARMCM33.S [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                .align   2
+
+                .thumb_func
+                .type    Reset_Handler, %function
+                .globl   Reset_Handler
+                .fnstart
+Reset_Handler:
+                ldr      r0, =__INITIAL_SP
+                msr      psp, r0
+
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM33/Source/ARM/startup_ARMCM33.S:86]: I believe there is some white space in this line. Would be good to remove necessary white spaces (maybe you can quickly check also after modified files).

File: Device/ARM/ARMCM35P/Source/ARM/ARMCM35P_ac6.sct [hunk 8]
```diff
@@ -32,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM35P/Source/ARM/ARMCM35P_ac6.sct:52]: I believe it should be 'boundary '

File: Device/ARM/ARMCM35P/Source/ARM/ARMCM35P_ac6.sct [hunk 9]
```diff
@@ -32,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
 
-#define __RW_BASE       __RAM_BASE
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM35P/Source/ARM/ARMCM35P_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM35P/Source/ARM/ARMCM35P_ac6_s.sct [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM35P/Source/ARM/ARMCM35P_ac6_s.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM55/Source/ARM/ARMCM55_ac6.sct [hunk 11]
```diff
@@ -40,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM55/Source/ARM/ARMCM55_ac6.sct:52]: I believe it should be 'boundary '

File: Device/ARM/ARMCM55/Source/ARM/ARMCM55_ac6.sct [hunk 12]
```diff
@@ -40,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
 
-#define __RW_BASE       __RAM_BASE
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM55/Source/ARM/ARMCM55_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMCM55/Source/ARM/ARMCM55_ac6_s.sct [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMCM55/Source/ARM/ARMCM55_ac6_s.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMv81MML/Source/ARM/ARMv81MML_ac6.sct [hunk 14]
```diff
@@ -40,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMv81MML/Source/ARM/ARMv81MML_ac6.sct:52]: I believe it should be 'boundary '

File: Device/ARM/ARMv81MML/Source/ARM/ARMv81MML_ac6.sct [hunk 15]
```diff
@@ -40,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
 
-#define __RW_BASE       __RAM_BASE
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMv81MML/Source/ARM/ARMv81MML_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMv81MML/Source/ARM/ARMv81MML_ac6_s.sct [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMv81MML/Source/ARM/ARMv81MML_ac6_s.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMv8MBL/Source/ARM/ARMv8MBL_ac6.sct [hunk 17]
```diff
@@ -32,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMv8MBL/Source/ARM/ARMv8MBL_ac6.sct:52]: I believe it should be 'boundary '

File: Device/ARM/ARMv8MBL/Source/ARM/ARMv8MBL_ac6.sct [hunk 18]
```diff
@@ -32,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
 
-#define __RW_BASE       __RAM_BASE
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMv8MBL/Source/ARM/ARMv8MBL_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMv8MBL/Source/ARM/ARMv8MBL_ac6_s.sct [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMv8MBL/Source/ARM/ARMv8MBL_ac6_s.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMv8MML/Source/ARM/ARMv8MML_ac6.sct [hunk 20]
```diff
@@ -32,33 +36,58 @@
...
+#define __CMSEVENEER_SIZE    0x200
+
 /*
 ;------------- <<< end of configuration section >>> ---------------------------
 */
 
 
 /*----------------------------------------------------------------------------
-  User Stack & Heap boundary definition
+  User Stack & Heap boundery definition
```
Comment (RobertRostohar) [inline Device/ARM/ARMv8MML/Source/ARM/ARMv8MML_ac6.sct:52]: I believe it should be 'boundary '

File: Device/ARM/ARMv8MML/Source/ARM/ARMv8MML_ac6.sct [hunk 21]
```diff
@@ -32,33 +36,58 @@
...
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
 
-#define __RW_BASE       __RAM_BASE
+#define __RW_BASE          ( __RAM_BASE )
 #define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMv8MML/Source/ARM/ARMv8MML_ac6.sct:81]: Maybe align it for better readability.

File: Device/ARM/ARMv8MML/Source/ARM/ARMv8MML_ac6_s.sct [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+#define __CV_SIZE          ( __CMSEVENEER_SIZE )
+#else
+#define __CV_SIZE          ( 0 )
+#endif
+
+#define __RO_BASE          ( __ROM_BASE )
+#define __RO_SIZE          ( __ROM_SIZE - __CV_SIZE )
+
+#define __RW_BASE          ( __RAM_BASE )
+#define __RW_SIZE      (__RAM_SIZE - __STACK_SIZE - __HEAP_SIZE)
```
Comment (RobertRostohar) [inline Device/ARM/ARMv8MML/Source/ARM/ARMv8MML_ac6_s.sct:81]: Maybe align it for better readability.

**Review by RobertRostohar — COMMENTED** [review]
Looks good to me. I haven't verified the latest version bu the concept was verified already earlier.

There are just a few white space and typo corrections that I have highlighted. Would be nice to update them before the templates get copied in user projects.

**Review by vovamarch — CHANGES_REQUESTED** [review]
As part of this commit we should also have some basic descriptions for CMSIS-Core documentation. 
 - Ref_Trustzone.txt for API references of  __TZ_set_STACKSEAL_S  and __STACK_SEAL define 
 - Template.txt mentioning stack sealing in startup file, with a reference to the upcoming appNote.

**Review by JonatanAntoni — APPROVED** [review]

File: CMSIS/Core/Include/cmsis_armclang.h [hunk 23]
```diff
@@ -136,6 +136,26 @@
...
+#endif
+
+#ifndef __TZ_STACK_SEAL_VALUE
+#define __TZ_STACK_SEAL_VALUE     0xFEF5EDA5FEF5EDA5ULL
+#endif
+
+
+__STATIC_FORCEINLINE void __TZ_set_STACKSEAL_S (uint32_t* stackTop) {
+  *((uint64_t *)stackTop) = __TZ_STACK_SEAL_VALUE;
+}
```
Comment (kjassmann-renesas) [inline CMSIS/Core/Include/cmsis_armclang.h:155]: Please add the same update to cmsis_iccarm.h

Comment (GuentherMartin) [reply CMSIS/Core/Include/cmsis_armclang.h:155]: Before we apply the changes to cmsis_iccarm.h, we have to clarify with IAR which linker generated symbol can be used.
In a first step I would add the changes to armclang and gcc and in a second step to iar.

Comment (JonatanAntoni) [reply CMSIS/Core/Include/cmsis_armclang.h:155]: cc: @TTornblom

**Review by vovamarch — APPROVED** [review]
Documentation changes look good.

**Comment by TTornblom** [comment]
I'm sorry that I haven't had the time to fix this for IAR yet.
Looking at the current implementation for ARM and GCC I can't help thinking that it appears overly complicated.
Wouldn't it be simpler to just push two words with the seal value on the stack in the startup, instead of creating separate areas in the linker scripts and storing the values there?
I also think that this could be done for all Cortex-M:s, not just the ones with Trust Zone support, unless they are extremely short on stack. That way they too would end up in the Hard Fault handler instead of executing undefined code.
What am I missing?

**Comment by JonatanAntoni** [comment]
Thanks, @TTornblom for coming back to this one.

With ARM and GCC, given the stacks are not deliberately placed into a NOINIT section, the memory will be overwritten during zero init. That's why the solution looks complicated. 

cc: @GuentherMartin, do you think it's worth to check the benefits of stack sealing used for all M-class devices?

To my experience stack underruns typically only happens if someone messes up the call stack, e.g. by a buffer overrun writing to the stack memory. In such situation its very likely that the sealing values get overwritten as well. To prevent such issues using the MPU sounds more adequate to me.

**Comment by TTornblom** [comment]
Thanks @JonatanAntoni,

Are the stacks always zero initialized for all Cortex-M devices with ARM and GCC?

I'm looking at the IAR startup and the stacks are not cleared on startup, and we changed the standard Cortex-M startup to always push the stack seal values, regardless of core type or security level. It wastes two words on the stack, and those who find this unacceptable can always add the startup to the project and comment out this.

If stack clearing is needed then we will have to rethink this.

**Comment by JonatanAntoni** [comment]
Hi @TTornblom,

With Arm Compiler (and GCC) we only reserve empty memory regions for stack and heap. They may or may-not be overwritten by C Library initialization code.

For stack sealing we added the provisions in this pull request. And the startup code templates make use of them. The C version of the startup code *does not* distingish the Compiler anymore (as the assembly version did)! Hence you need to provide a smiliar implementation for IAR for the function `__TZ_set_STACKSEAL_S(uint32_t *stack)` and the memory location `__STACK_SEAL` providing 8 bytes of memory right behind the stack memory area.

Is this a feasible solution for you?

Cheers,
Jonatan

**Pull request merged**